File: backend/src/main/java/com/hostel/backend/service/ComplaintService.java
package com.hostel.backend.service;

import com.hostel.backend.dto.*;
import com.hostel.backend.entity.Complaint;
import com.hostel.backend.entity.User;
import com.hostel.backend.enums.ComplaintStatus;
import com.hostel.backend.enums.Role;
import com.hostel.backend.exception.ResourceNotFoundException;
import com.hostel.backend.exception.UnauthorizedException;
import com.hostel.backend.repository.ComplaintRepository;
import com.hostel.backend.repository.UserRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class ComplaintService {

    private final ComplaintRepository      complaintRepository;
    private final UserRepository           userRepository;
    private final NotificationService      notificationService;
    private final ComplaintAuditLogService auditLogService;

    public ComplaintService(ComplaintRepository complaintRepository,
                            UserRepository userRepository,
                            NotificationService notificationService,
                            ComplaintAuditLogService auditLogService) {
        this.complaintRepository = complaintRepository;
        this.userRepository      = userRepository;
        this.notificationService = notificationService;
        this.auditLogService     = auditLogService;
    }

    // ─── Helpers ────────────────────────────────────────────────────────────

    private User getUser(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("User not found: " + email));
    }

    private Complaint getComplaint(Long id) {
        return complaintRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Complaint not found: " + id));
    }

    private boolean isStaff(Role role) {
        return role == Role.ADMIN || role == Role.WARDEN || role == Role.CARETAKER;
    }

    private UserSummaryDTO toUserSummary(User user) {
        if (user == null) return null;
        return UserSummaryDTO.builder()
                .id(user.getId())
                .name(user.getName())
                .email(user.getEmail())
                .role(user.getRole())
                .phoneNumber(user.getPhoneNumber())
                .hostelBlock(user.getHostelBlock())
                .roomNumber(user.getRoomNumber())
                .department(user.getDepartment())
                .build();
    }

    private ComplaintResponse toResponse(Complaint c) {
        return ComplaintResponse.builder()
                .id(c.getId())
                .title(c.getTitle())
                .description(c.getDescription())
                .category(c.getCategory())
                .priority(c.getPriority())
                .status(c.getStatus())
                .issuePhotoUrl(c.getIssuePhotoUrl())
                .resolvedPhotoUrl(c.getResolvedPhotoUrl())
                .rejectionReason(c.getRejectionReason())
                .rating(c.getRating())
                .overdue(c.isOverdue())
                .escalated(c.isEscalated())
                .student(toUserSummary(c.getStudent()))
                .assignedWorker(toUserSummary(c.getAssignedWorker()))
                .createdAt(c.getCreatedAt())
                .updatedAt(c.getUpdatedAt())
                .build();
    }

    // ─── Create ─────────────────────────────────────────────────────────────

    @Transactional
    public ComplaintResponse createComplaint(String email, ComplaintRequest request) {
        User user = getUser(email);
        if (user.getRole() != Role.STUDENT) {
            throw new UnauthorizedException("Only students can create complaints");
        }
        Complaint complaint = Complaint.builder()
                .title(request.getTitle())
                .category(request.getCategory())
                .priority(request.getPriority())
                .description(request.getDescription())
                .issuePhotoUrl(request.getImageUrl())
                .status(ComplaintStatus.CREATED)
                .student(user)
                .build();
        Complaint saved = complaintRepository.save(complaint);

        auditLogService.log(saved.getId(), user,
                null, ComplaintStatus.CREATED.name(),
                "Complaint filed by student");

        List<User> staff = userRepository.findByRole(Role.WARDEN);
        staff.addAll(userRepository.findByRole(Role.CARETAKER));
        staff.addAll(userRepository.findByRole(Role.ADMIN));
        staff.forEach(s -> notificationService.sendNotification(s,
                user.getName() + " filed a new complaint: " + saved.getTitle(),
                "COMPLAINT_CREATED"));

        return toResponse(saved);
    }

    // ─── Assign ─────────────────────────────────────────────────────────────

    @Transactional
    public ComplaintResponse assignWorker(Long complaintId, Long workerId, String email) {
        User assignedBy = getUser(email);
        if (!isStaff(assignedBy.getRole())) {
            throw new UnauthorizedException("Only admin, caretaker or warden can assign worker");
        }
        Complaint complaint = getComplaint(complaintId);
        User worker = userRepository.findById(workerId)
                .orElseThrow(() -> new ResourceNotFoundException("Worker not found: " + workerId));
        if (worker.getRole() != Role.WORKER) {
            throw new UnauthorizedException("Assigned user is not a worker");
        }

        String prevStatus = complaint.getStatus().name();
        complaint.setAssignedWorker(worker);
        complaint.setStatus(ComplaintStatus.ASSIGNED);
        Complaint saved = complaintRepository.save(complaint);

        auditLogService.log(saved.getId(), assignedBy,
                prevStatus, ComplaintStatus.ASSIGNED.name(),
                "Assigned to worker: " + worker.getName());

        notificationService.sendNotification(worker,
                "You have been assigned complaint: " + saved.getTitle(),
                "COMPLAINT_ASSIGNED");
        notificationService.sendNotification(saved.getStudent(),
                "Your complaint '" + saved.getTitle() + "' has been assigned to " + worker.getName(),
                "COMPLAINT_ASSIGNED");

        return toResponse(saved);
    }

    // ─── Reassign ───────────────────────────────────────────────────────────

    @Transactional
    public ComplaintResponse reassignWorker(Long complaintId, Long newWorkerId, String email) {
        User assignedBy = getUser(email);
        if (!isStaff(assignedBy.getRole())) {
            throw new UnauthorizedException("Only admin, caretaker or warden can reassign worker");
        }
        Complaint complaint = getComplaint(complaintId);
        User newWorker = userRepository.findById(newWorkerId)
                .orElseThrow(() -> new ResourceNotFoundException("Worker not found: " + newWorkerId));
        if (newWorker.getRole() != Role.WORKER) {
            throw new UnauthorizedException("Assigned user is not a worker");
        }

        User oldWorker    = complaint.getAssignedWorker();
        String prevStatus = complaint.getStatus().name();
        complaint.setAssignedWorker(newWorker);
        complaint.setStatus(ComplaintStatus.ASSIGNED);
        Complaint saved = complaintRepository.save(complaint);

        auditLogService.log(saved.getId(), assignedBy,
                prevStatus, ComplaintStatus.ASSIGNED.name(),
                "Reassigned from " +
                (oldWorker != null ? oldWorker.getName() : "nobody") +
                " to " + newWorker.getName());

        if (oldWorker != null) {
            notificationService.sendNotification(oldWorker,
                    "Complaint '" + saved.getTitle() + "' has been reassigned from you",
                    "COMPLAINT_ASSIGNED");
        }
        notificationService.sendNotification(newWorker,
                "Complaint '" + saved.getTitle() + "' has been reassigned to you",
                "COMPLAINT_ASSIGNED");
        notificationService.sendNotification(saved.getStudent(),
                "Your complaint '" + saved.getTitle() + "' reassigned to " + newWorker.getName(),
                "COMPLAINT_ASSIGNED");

        return toResponse(saved);
    }

    // ─── Reject ─────────────────────────────────────────────────────────────

    @Transactional
    public ComplaintResponse rejectComplaint(Long complaintId, String reason, String email) {
        User rejectedBy = getUser(email);
        if (!isStaff(rejectedBy.getRole())) {
            throw new UnauthorizedException("Only admin, caretaker or warden can reject complaints");
        }
        if (reason == null || reason.trim().isEmpty()) {
            throw new RuntimeException("Rejection reason is required");
        }
        Complaint complaint = getComplaint(complaintId);
        if (complaint.getStatus() == ComplaintStatus.CLOSED) {
            throw new RuntimeException("Cannot reject a closed complaint");
        }

        String prevStatus = complaint.getStatus().name();
        complaint.setStatus(ComplaintStatus.REJECTED);
        complaint.setRejectionReason(reason.trim());
        Complaint saved = complaintRepository.save(complaint);

        auditLogService.log(saved.getId(), rejectedBy,
                prevStatus, ComplaintStatus.REJECTED.name(),
                "Rejected. Reason: " + reason.trim());

        notificationService.sendNotification(saved.getStudent(),
                "Your complaint '" + saved.getTitle() + "' was rejected. Reason: " + reason,
                "STATUS_UPDATED");

        return toResponse(saved);
    }

    // ─── Update Status (worker) ──────────────────────────────────────────────

    @Transactional
    public ComplaintResponse updateStatus(Long complaintId,
                                          UpdateComplaintStatusRequest request,
                                          String email) {
        User worker = getUser(email);
        if (worker.getRole() != Role.WORKER) {
            throw new UnauthorizedException("Only worker can update complaint status");
        }
        Complaint complaint = getComplaint(complaintId);
        if (complaint.getAssignedWorker() == null ||
                !complaint.getAssignedWorker().getId().equals(worker.getId())) {
            throw new UnauthorizedException("You are not assigned to this complaint");
        }

        String prevStatus = complaint.getStatus().name();
        complaint.setStatus(request.getStatus());
        if (request.getResolvedPhotoUrl() != null) {
            complaint.setResolvedPhotoUrl(request.getResolvedPhotoUrl());
        }
        Complaint saved = complaintRepository.save(complaint);

        auditLogService.log(saved.getId(), worker,
                prevStatus, request.getStatus().name(),
                request.getStatus() == ComplaintStatus.IN_PROGRESS
                        ? "Worker started working"
                        : "Worker marked as resolved");

        if (request.getStatus() == ComplaintStatus.IN_PROGRESS) {
            notificationService.sendNotification(saved.getStudent(),
                    "Work has started on your complaint: " + saved.getTitle(),
                    "STATUS_UPDATED");
        } else if (request.getStatus() == ComplaintStatus.RESOLVED) {
            notificationService.sendNotification(saved.getStudent(),
                    "Your complaint '" + saved.getTitle() + "' is resolved. Please close it.",
                    "STATUS_UPDATED");
        }

        return toResponse(saved);
    }

    // ─── Close + Rate ────────────────────────────────────────────────────────

    @Transactional
    public ComplaintResponse closeComplaint(Long complaintId,
                                            CloseComplaintRequest request,
                                            String email) {
        User student = getUser(email);
        if (student.getRole() != Role.STUDENT) {
            throw new UnauthorizedException("Only student can close complaint");
        }
        Complaint complaint = getComplaint(complaintId);
        if (!complaint.getStudent().getId().equals(student.getId())) {
            throw new UnauthorizedException("You cannot close this complaint");
        }
        if (complaint.getStatus() != ComplaintStatus.RESOLVED) {
            throw new RuntimeException("Complaint must be RESOLVED before closing");
        }

        if (request != null && request.getRating() != null) {
            int r = request.getRating();
            if (r < 1 || r > 5) throw new RuntimeException("Rating must be between 1 and 5");
            complaint.setRating(r);
        }

        complaint.setStatus(ComplaintStatus.CLOSED);
        Complaint saved = complaintRepository.save(complaint);

        auditLogService.log(saved.getId(), student,
                ComplaintStatus.RESOLVED.name(), ComplaintStatus.CLOSED.name(),
                "Closed by student" + (saved.getRating() != null
                        ? " | Rating: " + saved.getRating() + "/5" : ""));

        return toResponse(saved);
    }

    // ─── Queries ────────────────────────────────────────────────────────────

    @Transactional(readOnly = true)
    public Page<ComplaintResponse> getComplaintsByRole(String email, int page,
                                                        int size, ComplaintStatus status) {
        User user = getUser(email);
        Pageable pageable = PageRequest.of(page, size);
        return switch (user.getRole()) {
            case STUDENT -> status != null
                    ? complaintRepository.findByStatusAndStudent(status, user, pageable).map(this::toResponse)
                    : complaintRepository.findByStudent(user, pageable).map(this::toResponse);
            case WORKER -> status != null
                    ? complaintRepository.findByAssignedWorkerAndStatus(user, status, pageable).map(this::toResponse)
                    : complaintRepository.findByAssignedWorker(user, pageable).map(this::toResponse);
            case ADMIN, CARETAKER, WARDEN -> status != null
                    ? complaintRepository.findByStatus(status, pageable).map(this::toResponse)
                    : complaintRepository.findAll(pageable).map(this::toResponse);
            default -> throw new UnauthorizedException("Invalid role for this operation");
        };
    }

    @Transactional(readOnly = true)
    public StudentDashboardStatsResponse getStudentDashboardStats(String email) {
        User user = getUser(email);
        if (user.getRole() != Role.STUDENT) {
            throw new UnauthorizedException("Access denied");
        }
        return new StudentDashboardStatsResponse(
                complaintRepository.countByStudent(user),
                complaintRepository.countByStudentAndStatus(user, ComplaintStatus.CREATED)
                + complaintRepository.countByStudentAndStatus(user, ComplaintStatus.ASSIGNED)
                + complaintRepository.countByStudentAndStatus(user, ComplaintStatus.IN_PROGRESS),
                complaintRepository.countByStudentAndStatus(user, ComplaintStatus.RESOLVED),
                complaintRepository.countByStudentAndStatus(user, ComplaintStatus.CLOSED),
                complaintRepository.countByStudentAndStatus(user, ComplaintStatus.REJECTED)
        );
    }

    // ✅ Worker dashboard stats — uses @Query AVG, no list fetch
    @Transactional(readOnly = true)
    public WorkerDashboardStatsResponse getWorkerDashboardStats(String email) {
        User worker = getUser(email);
        if (worker.getRole() != Role.WORKER) {
            throw new UnauthorizedException("Access denied");
        }

        long assigned   = complaintRepository.countByAssignedWorkerAndStatus(worker, ComplaintStatus.ASSIGNED);
        long inProgress = complaintRepository.countByAssignedWorkerAndStatus(worker, ComplaintStatus.IN_PROGRESS);
        long resolved   = complaintRepository.countByAssignedWorkerAndStatus(worker, ComplaintStatus.RESOLVED);
        long closed     = complaintRepository.countByAssignedWorkerAndStatus(worker, ComplaintStatus.CLOSED);

        // ✅ Fixed — use @Query method, no list fetching
        Double raw = complaintRepository.findAverageRatingByWorker(worker);
        Double averageRating = raw != null ? Math.round(raw * 10.0) / 10.0 : null;

        return new WorkerDashboardStatsResponse(assigned, inProgress, resolved, closed, averageRating);
    }

    @Transactional(readOnly = true)
    public DashboardStatsResponse getDashboardStats(String email) {
        User user = getUser(email);
        if (!isStaff(user.getRole())) {
            throw new UnauthorizedException("Access denied");
        }
        return new DashboardStatsResponse(
                complaintRepository.count(),
                complaintRepository.countByStatus(ComplaintStatus.CREATED),
                complaintRepository.countByStatus(ComplaintStatus.ASSIGNED),
                complaintRepository.countByStatus(ComplaintStatus.IN_PROGRESS),
                complaintRepository.countByStatus(ComplaintStatus.RESOLVED),
                complaintRepository.countByStatus(ComplaintStatus.CLOSED),
                complaintRepository.countByStatus(ComplaintStatus.REJECTED)
        );
    }

    @Transactional(readOnly = true)
    public ComplaintResponse getComplaintById(Long id, String email) {
        User user = getUser(email);
        Complaint complaint = getComplaint(id);
        if (user.getRole() == Role.STUDENT &&
                !complaint.getStudent().getId().equals(user.getId())) {
            throw new UnauthorizedException("Access denied");
        }
        return toResponse(complaint);
    }

    // ─── Worker Stats by Admin ───────────────────────────────────────────────

    @Transactional(readOnly = true)
    public WorkerDashboardStatsResponse getWorkerStatsByAdmin(Long workerId) {
        User worker = userRepository.findById(workerId)
                .orElseThrow(() -> new ResourceNotFoundException("Worker not found"));

        long assigned   = complaintRepository.countByAssignedWorkerAndStatus(worker, ComplaintStatus.ASSIGNED);
        long inProgress = complaintRepository.countByAssignedWorkerAndStatus(worker, ComplaintStatus.IN_PROGRESS);
        long resolved   = complaintRepository.countByAssignedWorkerAndStatus(worker, ComplaintStatus.RESOLVED);
        long closed     = complaintRepository.countByAssignedWorkerAndStatus(worker, ComplaintStatus.CLOSED);

        Double raw = complaintRepository.findAverageRatingByWorker(worker);
        Double averageRating = raw != null ? Math.round(raw * 10.0) / 10.0 : null;

        return new WorkerDashboardStatsResponse(assigned, inProgress, resolved, closed, averageRating);
    }
}

File: backend/src/services/complaint.service.js
const prisma = require('../config/prisma');
const { NotFoundError, ForbiddenError, BadRequestError } = require('../utils/AppError');
const { toUserSummary } = require('../utils/mappers');
const { toPage } = require('../utils/pagination');
const auditLogService = require('./complaintAuditLog.service');
const notificationService = require('./notification.service');

const STAFF_ROLES = ['ADMIN', 'WARDEN', 'CARETAKER'];
const isStaff = (role) => STAFF_ROLES.includes(role);

// `block` is a plain enum column on User now, not a relation — only
// `department` needs an include. See docs/MIGRATION_NOTES.md "Hostel-10 scoping".
const personInclude = {
  department: true,
};

const complaintInclude = {
  student: { include: personInclude },
  assignedWorker: { include: personInclude },
};

async function getUserByEmail(email) {
  const user = await prisma.user.findUnique({ where: { email } });
  if (!user) throw new NotFoundError(`User not found: ${email}`);
  return user;
}

async function getComplaintOrThrow(id, include = complaintInclude) {
  const complaint = await prisma.complaint.findUnique({ where: { id: BigInt(id) }, include });
  if (!complaint) throw new NotFoundError(`Complaint not found: ${id}`);
  return complaint;
}

async function toResponse(complaint) {
  const media = await prisma.complaintMedia.findMany({ where: { complaintId: complaint.id } });
  const mediaUrls = media.filter((m) => m.uploadedBy === 'STUDENT').map((m) => m.mediaUrl);
  const resolvedMediaUrls = media.filter((m) => m.uploadedBy === 'WORKER').map((m) => m.mediaUrl);

  return {
    id: complaint.id,
    title: complaint.title,
    description: complaint.description,
    category: complaint.category,
    priority: complaint.priority,
    status: complaint.status,
    mediaUrls,
    resolvedMediaUrls,
    rejectionReason: complaint.rejectionReason,
    rating: complaint.rating,
    reviewText: complaint.reviewText,
    disputeReason: complaint.disputeReason,
    disputedAt: complaint.disputedAt,
    overdue: complaint.isOverdue,
    escalated: complaint.isEscalated,
    slot1Day: complaint.slot1Day,
    slot1Time: complaint.slot1Time,
    slot2Day: complaint.slot2Day,
    slot2Time: complaint.slot2Time,
    slot3Day: complaint.slot3Day,
    slot3Time: complaint.slot3Time,
    student: toUserSummary(complaint.student),
    assignedWorker: toUserSummary(complaint.assignedWorker),
    createdAt: complaint.createdAt,
    updatedAt: complaint.updatedAt,
  };
}

// ─── Create ─────────────────────────────────────────────────────────────────

async function createComplaint(email, request) {
  const user = await getUserByEmail(email);
  if (user.role !== 'STUDENT') throw new ForbiddenError('Only students can create complaints');
  if (request.mediaUrls && request.mediaUrls.length > 3) {
    throw new BadRequestError('Maximum 3 photos allowed');
  }

  const saved = await prisma.complaint.create({
    data: {
      title: request.title,
      category: request.category,
      priority: request.priority || 'LOW',
      description: request.description,
      status: 'CREATED',
      block: user.block,
      studentId: user.id,
      slot1Day: request.slot1Day,
      slot1Time: request.slot1Time,
      slot2Day: request.slot2Day,
      slot2Time: request.slot2Time,
      slot3Day: request.slot3Day,
      slot3Time: request.slot3Time,
      isOverdue: false,
      isEscalated: false,
    },
    include: complaintInclude,
  });

  const mediaUrls = (request.mediaUrls || []).filter((u) => u && u.trim());
  if (mediaUrls.length) {
    await prisma.complaintMedia.createMany({
      data: mediaUrls.map((mediaUrl) => ({
        complaintId: saved.id,
        mediaUrl,
        uploadedBy: 'STUDENT',
      })),
    });
  }

  await auditLogService.log(saved.id, user, null, 'CREATED', 'Complaint filed by student');

  const staff = await prisma.user.findMany({ where: { role: { in: ['WARDEN', 'CARETAKER', 'ADMIN'] } } });
  await Promise.all(
    staff.map((s) =>
      notificationService.sendNotification(s, `${user.name} filed a new complaint: ${saved.title}`, 'COMPLAINT_CREATED')
    )
  );

  return toResponse(saved);
}

// ─── Assign ─────────────────────────────────────────────────────────────────

async function assignWorker(complaintId, workerId, email) {
  const assignedBy = await getUserByEmail(email);
  if (!isStaff(assignedBy.role)) throw new ForbiddenError('Only admin, caretaker or warden can assign worker');

  const complaint = await getComplaintOrThrow(complaintId);
  const worker = await prisma.user.findUnique({ where: { id: BigInt(workerId) } });
  if (!worker) throw new NotFoundError(`Worker not found: ${workerId}`);
  if (worker.role !== 'WORKER') throw new ForbiddenError('Assigned user is not a worker');

  const prevStatus = complaint.status;
  const saved = await prisma.complaint.update({
    where: { id: complaint.id },
    data: { workerId: worker.id, status: 'ASSIGNED' },
    include: complaintInclude,
  });

  await auditLogService.log(saved.id, assignedBy, prevStatus, 'ASSIGNED', `Assigned to worker: ${worker.name}`);

  await notificationService.sendNotification(worker, `You have been assigned complaint: ${saved.title}`, 'COMPLAINT_ASSIGNED');
  await notificationService.sendNotification(
    saved.student,
    `Your complaint '${saved.title}' has been assigned to ${worker.name}`,
    'COMPLAINT_ASSIGNED'
  );

  return toResponse(saved);
}

// ─── Reassign ───────────────────────────────────────────────────────────────

async function reassignWorker(complaintId, newWorkerId, email) {
  const assignedBy = await getUserByEmail(email);
  if (!isStaff(assignedBy.role)) throw new ForbiddenError('Only admin, caretaker or warden can reassign worker');

  const complaint = await getComplaintOrThrow(complaintId);
  const newWorker = await prisma.user.findUnique({ where: { id: BigInt(newWorkerId) } });
  if (!newWorker) throw new NotFoundError(`Worker not found: ${newWorkerId}`);
  if (newWorker.role !== 'WORKER') throw new ForbiddenError('Assigned user is not a worker');

  const oldWorker = complaint.assignedWorker;
  const prevStatus = complaint.status;

  const saved = await prisma.complaint.update({
    where: { id: complaint.id },
    data: { workerId: newWorker.id, status: 'ASSIGNED' },
    include: complaintInclude,
  });

  await auditLogService.log(
    saved.id,
    assignedBy,
    prevStatus,
    'ASSIGNED',
    `Reassigned from ${oldWorker ? oldWorker.name : 'nobody'} to ${newWorker.name}`
  );

  if (oldWorker) {
    await notificationService.sendNotification(oldWorker, `Complaint '${saved.title}' has been reassigned from you`, 'COMPLAINT_ASSIGNED');
  }
  await notificationService.sendNotification(newWorker, `Complaint '${saved.title}' has been reassigned to you`, 'COMPLAINT_ASSIGNED');
  await notificationService.sendNotification(
    saved.student,
    `Your complaint '${saved.title}' reassigned to ${newWorker.name}`,
    'COMPLAINT_ASSIGNED'
  );

  return toResponse(saved);
}

// ─── Reject ─────────────────────────────────────────────────────────────────

async function rejectComplaint(complaintId, reason, email) {
  const rejectedBy = await getUserByEmail(email);
  if (!isStaff(rejectedBy.role)) throw new ForbiddenError('Only admin, caretaker or warden can reject complaints');
  if (!reason || !reason.trim()) throw new BadRequestError('Rejection reason is required');

  const complaint = await getComplaintOrThrow(complaintId);
  if (complaint.status === 'CLOSED') throw new BadRequestError('Cannot reject a closed complaint');

  const prevStatus = complaint.status;
  const saved = await prisma.complaint.update({
    where: { id: complaint.id },
    data: { status: 'REJECTED', rejectionReason: reason.trim() },
    include: complaintInclude,
  });

  await auditLogService.log(saved.id, rejectedBy, prevStatus, 'REJECTED', `Rejected. Reason: ${reason.trim()}`);
  await notificationService.sendNotification(
    saved.student,
    `Your complaint '${saved.title}' was rejected. Reason: ${reason}`,
    'STATUS_UPDATED'
  );

  return toResponse(saved);
}

// ─── Update status (worker) ─────────────────────────────────────────────────

async function updateStatus(complaintId, request, email) {
  const worker = await getUserByEmail(email);
  if (worker.role !== 'WORKER') throw new ForbiddenError('Only worker can update complaint status');

  const complaint = await getComplaintOrThrow(complaintId);
  if (!complaint.workerId || complaint.workerId !== worker.id) {
    throw new ForbiddenError('You are not assigned to this complaint');
  }

  const prevStatus = complaint.status;
  const data = { status: request.status };

  if (request.status === 'RESOLVED' && request.resolvedMediaUrls && request.resolvedMediaUrls.length) {
    if (request.resolvedMediaUrls.length > 3) throw new BadRequestError('Maximum 3 proof photos allowed');

    await prisma.complaintMedia.deleteMany({
      where: { complaintId: complaint.id, uploadedBy: 'WORKER' },
    });

    const urls = request.resolvedMediaUrls.filter((u) => u && u.trim());
    if (urls.length) {
      await prisma.complaintMedia.createMany({
        data: urls.map((mediaUrl) => ({ complaintId: complaint.id, mediaUrl, uploadedBy: 'WORKER' })),
      });
    }
  }

  // NOTE: the original ComplaintService.updateStatus() never sets
  // `resolvedAt` even though the column exists (it's used only by two
  // unused AnalyticsService repository queries) — left unset here too, for
  // exact parity. See docs/MIGRATION_NOTES.md.
  const saved = await prisma.complaint.update({ where: { id: complaint.id }, data, include: complaintInclude });

  await auditLogService.log(
    saved.id,
    worker,
    prevStatus,
    request.status,
    request.status === 'IN_PROGRESS' ? 'Worker started working' : 'Worker marked as resolved'
  );

  if (request.status === 'IN_PROGRESS') {
    await notificationService.sendNotification(saved.student, `Work has started on your complaint: ${saved.title}`, 'STATUS_UPDATED');
  } else if (request.status === 'RESOLVED') {
    await notificationService.sendNotification(
      saved.student,
      `Your complaint '${saved.title}' is resolved. Please close it or flag if still unresolved.`,
      'STATUS_UPDATED'
    );
  }

  return toResponse(saved);
}

// ─── Close ──────────────────────────────────────────────────────────────────

async function closeComplaint(complaintId, request, email) {
  const student = await getUserByEmail(email);
  if (student.role !== 'STUDENT') throw new ForbiddenError('Only student can close complaint');

  const complaint = await getComplaintOrThrow(complaintId);
  if (complaint.studentId !== student.id) throw new ForbiddenError('You cannot close this complaint');
  if (complaint.status !== 'RESOLVED') throw new BadRequestError('Complaint must be RESOLVED before closing');

  const data = { status: 'CLOSED' };
  if (request) {
    if (request.rating != null) {
      const r = request.rating;
      if (r < 1 || r > 5) throw new BadRequestError('Rating must be between 1 and 5');
      data.rating = r;
    }
    if (request.reviewText && request.reviewText.trim()) {
      data.reviewText = request.reviewText.trim();
    }
  }

  const saved = await prisma.complaint.update({ where: { id: complaint.id }, data, include: complaintInclude });

  await auditLogService.log(
    saved.id,
    student,
    'RESOLVED',
    'CLOSED',
    'Closed by student' + (saved.rating != null ? ` | Rating: ${saved.rating}/5` : '') + (saved.reviewText ? ' | Review given' : '')
  );

  if (saved.assignedWorker) {
    const ratingText = saved.rating != null ? ` (Rated ${saved.rating}/5)` : '';
    await notificationService.sendNotification(
      saved.assignedWorker,
      `Complaint '${saved.title}' was closed by the student${ratingText}`,
      'COMPLAINT_CLOSED'
    );
  }

  return toResponse(saved);
}

// ─── Dispute ────────────────────────────────────────────────────────────────

async function disputeComplaint(complaintId, request, email) {
  const student = await getUserByEmail(email);
  if (student.role !== 'STUDENT') throw new ForbiddenError('Only student can dispute a complaint');

  const complaint = await getComplaintOrThrow(complaintId);
  if (complaint.studentId !== student.id) throw new ForbiddenError('You cannot dispute this complaint');
  if (complaint.status !== 'RESOLVED') throw new BadRequestError('You can only dispute a RESOLVED complaint');
  if (!request.disputeReason || !request.disputeReason.trim()) {
    throw new BadRequestError('Please provide a reason for disputing');
  }

  const reason = request.disputeReason.trim();
  const saved = await prisma.complaint.update({
    where: { id: complaint.id },
    data: { status: 'DISPUTED', disputeReason: reason, disputedAt: new Date() },
    include: complaintInclude,
  });

  await auditLogService.log(saved.id, student, 'RESOLVED', 'DISPUTED', `Student disputed: ${reason}`);

  const toNotify = await prisma.user.findMany({ where: { role: { in: ['CARETAKER', 'WARDEN'] } } });
  if (saved.assignedWorker) toNotify.push(saved.assignedWorker);
  await Promise.all(
    toNotify.map((u) =>
      notificationService.sendNotification(u, `⚠️ Student disputed complaint '${saved.title}'. Reason: ${reason}`, 'COMPLAINT_DISPUTED')
    )
  );

  return toResponse(saved);
}

// ─── Queries ────────────────────────────────────────────────────────────────

function buildDateRange(dateFrom, dateTo) {
  const range = {};
  if (dateFrom) range.gte = new Date(`${dateFrom}T00:00:00`);
  if (dateTo) range.lte = new Date(`${dateTo}T23:59:59`);
  return Object.keys(range).length ? range : undefined;
}

async function getComplaintsByRole(email, page, size, filters) {
  const { status, blockName, category, priority, dateFrom, dateTo } = filters;
  const user = await getUserByEmail(email);

  const where = {};
  if (status) where.status = status;
  if (category) where.category = category;
  if (priority) where.priority = priority;
  const createdAt = buildDateRange(dateFrom, dateTo);
  if (createdAt) where.createdAt = createdAt;

  if (user.role === 'STUDENT') {
    where.studentId = user.id;
  } else if (user.role === 'WORKER') {
    where.workerId = user.id;
  } else if (isStaff(user.role)) {
    // `blockName` kept as the query param name for API compatibility; its
    // value is now just the Block enum ('A'/'B') rather than a joined
    // Block.name — see docs/MIGRATION_NOTES.md "Hostel-10 scoping".
    if (blockName) where.block = blockName;
  } else {
    throw new ForbiddenError('Invalid role for this operation');
  }

  const [rows, total] = await Promise.all([
    prisma.complaint.findMany({
      where,
      include: complaintInclude,
      orderBy: { createdAt: 'desc' },
      skip: page * size,
      take: size,
    }),
    prisma.complaint.count({ where }),
  ]);

  const content = await Promise.all(rows.map(toResponse));
  return toPage(content, total, page, size);
}

async function getStudentDashboardStats(email) {
  const user = await getUserByEmail(email);
  if (user.role !== 'STUDENT') throw new ForbiddenError('Access denied');

  const [total, created, assigned, inProgress, resolved, closed, rejected, disputed] = await Promise.all([
    prisma.complaint.count({ where: { studentId: user.id } }),
    prisma.complaint.count({ where: { studentId: user.id, status: 'CREATED' } }),
    prisma.complaint.count({ where: { studentId: user.id, status: 'ASSIGNED' } }),
    prisma.complaint.count({ where: { studentId: user.id, status: 'IN_PROGRESS' } }),
    prisma.complaint.count({ where: { studentId: user.id, status: 'RESOLVED' } }),
    prisma.complaint.count({ where: { studentId: user.id, status: 'CLOSED' } }),
    prisma.complaint.count({ where: { studentId: user.id, status: 'REJECTED' } }),
    prisma.complaint.count({ where: { studentId: user.id, status: 'DISPUTED' } }),
  ]);

  return {
    total,
    pending: created + assigned + inProgress,
    resolved,
    closed,
    rejected,
    disputed,
  };
}

async function averageRatingForWorker(workerId) {
  const result = await prisma.complaint.aggregate({
    where: { workerId, rating: { not: null } },
    _avg: { rating: true },
  });
  return result._avg.rating;
}

async function workerDashboardStats(workerId) {
  const [assigned, inProgress, resolved, closed, avg] = await Promise.all([
    prisma.complaint.count({ where: { workerId, status: 'ASSIGNED' } }),
    prisma.complaint.count({ where: { workerId, status: 'IN_PROGRESS' } }),
    prisma.complaint.count({ where: { workerId, status: 'RESOLVED' } }),
    prisma.complaint.count({ where: { workerId, status: 'CLOSED' } }),
    averageRatingForWorker(workerId),
  ]);
  return {
    assigned,
    inProgress,
    resolved,
    closed,
    averageRating: avg != null ? Math.round(avg * 10) / 10 : null,
  };
}

async function getWorkerDashboardStats(email) {
  const worker = await getUserByEmail(email);
  if (worker.role !== 'WORKER') throw new ForbiddenError('Access denied');
  return workerDashboardStats(worker.id);
}

async function getWorkerStatsByAdmin(workerId) {
  const worker = await prisma.user.findUnique({ where: { id: BigInt(workerId) } });
  if (!worker) throw new NotFoundError('Worker not found');
  return workerDashboardStats(worker.id);
}

async function getDashboardStats(email) {
  const user = await getUserByEmail(email);
  if (!isStaff(user.role)) throw new ForbiddenError('Access denied');

  const [total, created, assigned, inProgress, resolved, disputed, closed, rejected] = await Promise.all([
    prisma.complaint.count(),
    prisma.complaint.count({ where: { status: 'CREATED' } }),
    prisma.complaint.count({ where: { status: 'ASSIGNED' } }),
    prisma.complaint.count({ where: { status: 'IN_PROGRESS' } }),
    prisma.complaint.count({ where: { status: 'RESOLVED' } }),
    prisma.complaint.count({ where: { status: 'DISPUTED' } }),
    prisma.complaint.count({ where: { status: 'CLOSED' } }),
    prisma.complaint.count({ where: { status: 'REJECTED' } }),
  ]);

  return { total, created, assigned, inProgress, resolved, disputed, closed, rejected };
}

async function getComplaintById(id, email) {
  const user = await getUserByEmail(email);
  const complaint = await getComplaintOrThrow(id);
  if (user.role === 'STUDENT' && complaint.studentId !== user.id) {
    throw new ForbiddenError('Access denied');
  }
  return toResponse(complaint);
}

module.exports = {
  createComplaint,
  assignWorker,
  reassignWorker,
  rejectComplaint,
  updateStatus,
  closeComplaint,
  disputeComplaint,
  getComplaintsByRole,
  getStudentDashboardStats,
  getWorkerDashboardStats,
  getDashboardStats,
  getComplaintById,
  getWorkerStatsByAdmin,
  isStaff,
};
